### PR TITLE
Add LEVIR-CD+ Dataset

### LINKED ISSUE #105: LEVIR-CD+ Dataset
The LEVIR-CD+ dataset, proposed in ["S2Looking: A Satellite Side-Looking Dataset for Building Change Detection", Shen et al.](https://arxiv.org/abs/2107.09244) is an urban building change detection dataset of 985 1024x1024 0.5m RGB image pairs extracted from Google Earth. The dataset contains building/land use change masks from 20 different regions of Texas between 2002-2020 with a time span of 5 years.

### PR DESCRIPTION
**Pull request by isaaccorley**
- Adds `torchgeo.datasets.LEVIRCDPlus`
- Updated docs
- Added `tests/datasets/test_levircd.py` (currently passing)
- Added sample data for tests `tests/data/levircd/LEVIR-CD.zip`

TODO:
- Some bug in `torchvision.datasets.utils.download_file_from_google_drive` is just downloading an empty file when using the dataset with `download=True`

Open to feedback and suggestions.

Closes #105

### CODE REVIEW

File: torchgeo/datasets/levircd.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+import torch
+from PIL import Image
+from torch import Tensor
+
+from .geo import VisionDataset
+from .utils import download_and_extract_archive
+
+
+class LEVIRCDPlus(VisionDataset):
+    """LEVIR-CD+ dataset.
```
Comment (adamjstewart) [inline torchgeo/datasets/levircd.py:20]: I think pydocstyle will ask for a blank line after the first line

File: torchgeo/datasets/levircd.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    def __len__(self) -> int:
+        """Return the number of data points in the dataset.
+
+        Returns:
+            length of the dataset
+        """
+        return len(self.files)
+
+    def _load_files(self, root: str, split: str) -> Dict[str, str]:
```
Comment (adamjstewart) [inline torchgeo/datasets/levircd.py:118]: Can you add a docstring for this function? It won't make it into the documentation, but it's helpful if anyone ever needs to edit this function.

File: torchgeo/datasets/levircd.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    * images are three-channel pngs
+    * masks are single-channel pngs where no change = 0, change = 255
+
+    Dataset classes:
+    1. no change
+    2. change
+
+    If you use this dataset in your research, please cite the following paper:
+    * https://arxiv.org/abs/2107.09244
+    """  # noqa: E501
```
Comment (adamjstewart) [inline torchgeo/datasets/levircd.py:40]: I think we can remove `# noqa: E501` here, none of the lines exceed 88 characters right?

File: torchgeo/datasets/levircd.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+    url = "https://drive.google.com/file/d/1JamSsxiytXdzAIk6VDVWfc-OsX-81U81/view?usp=sharing"
```
Comment (adamjstewart) [inline torchgeo/datasets/levircd.py:42]: If you remove `/view?usp=sharing` does that help with the download? The `torchgeo.datasets.nwpu.VHR10` dataset downloads from Google Drive successfully so I don't _think_ anything is broken.

File: torchgeo/datasets/levircd.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    md5 = "1adf156f628aa32fb2e8fe6cada16c04"
+    filename = "LEVIR-CD+.zip"
+    filenames = ["train", "test"]
```
Comment (adamjstewart) [inline torchgeo/datasets/levircd.py:45]: I would call this `splits` to help disambiguate from `filename`.

File: torchgeo/datasets/levircd.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+                entry and returns a transformed version
+            download: if True, download dataset and store it in the root directory
+            checksum: if True, check the MD5 of the downloaded files (may be slow)
+
+        Raises:
+            AssertionError: if ``split`` argument is invalid
+            RuntimeError: if ``download=False`` and data is not found, or checksums
+                don't match
+        """
+        assert split in ["train", "test"]
```
Comment (adamjstewart) [inline torchgeo/datasets/levircd.py:70]: You could use `self.splits` here (after the change suggested above) to prevent code duplication

**Review by adamjstewart — CHANGES_REQUESTED** [review]
This looks great, @calebrob6 and I were just talking about how we could really use a change detection dataset.

Thanks for adding unit tests too! For the test data, we're trying to avoid including real data from the original dataset because of copyright/licensing issues, and also due to file size. Can you create fake PNG files (as small as possible)? See https://github.com/microsoft/torchgeo/tree/main/tests/data#visiondataset for instructions on how the other test data was created.

Other than that, just a few very minor suggested changes.

File: torchgeo/datasets/levircd.py (hunk 1 above)
Comment (isaaccorley) [reply torchgeo/datasets/levircd.py:20]: Fixed.

File: torchgeo/datasets/levircd.py (hunk 2 above)
Comment (isaaccorley) [reply torchgeo/datasets/levircd.py]: Added docstring.

File: torchgeo/datasets/levircd.py (hunk 3 above)
Comment (isaaccorley) [reply torchgeo/datasets/levircd.py]: Removed

File: torchgeo/datasets/levircd.py (hunk 4 above)
Comment (isaaccorley) [reply torchgeo/datasets/levircd.py]: I actually just installed torchgeo into a Colab instance and tried instantiating the VHR10 dataset with `download=True` and it gave me the same issue where the file is empty so it throws an error for not being a .rar file. Not sure what's going on.

File: torchgeo/datasets/levircd.py (hunk 5 above)
Comment (isaaccorley) [reply torchgeo/datasets/levircd.py]: Renamed to `splits`

File: torchgeo/datasets/levircd.py (hunk 6 above)
Comment (isaaccorley) [reply torchgeo/datasets/levircd.py]: Changed `["train", "test"]` to `self.splits`

**Comment by isaaccorley** [comment]
Something to note is that for the pair of images, I concatenated them on the channel dim, `(2c, h, w)` but maybe it would be better to stack them to `(2, c, h, w)`. Any thoughts?

I'm mainly asking because I have quite a few other change detection datasets I could add and want to make the outputs uniform across change detection datasets.

File: torchgeo/datasets/levircd.py (hunk 4 above)
Comment (adamjstewart) [reply torchgeo/datasets/levircd.py]: Huh, you're right, I get the same thing. Looks like things must have broken. I'll submit a PR to try to fix VHR10 and if it works I'll let you know what to change.

**Comment by adamjstewart** [comment]
For the channel question, @calebrob6 had a very similar question when he was working on the `ChesapeakeCVPR` dataset, where each dataset contains 2 Sentinel images. In that case, we decided to concatenate them along the channel dimension because most users are likely to pass a single image stack to a model.

For change detection, do people usually pass a single image, or two separate images?

**Comment by isaaccorley** [comment]
In other implementations, I've seen it passed through like `model(x1, x2)` or with a temporal dim like `(b, t, c, h, w)` where `t=2` and passing to the model `model(x)`. In the [change detection models](https://github.com/isaaccorley/torchrs/blob/main/torchrs/models/fc_cd.py) I wrote for torchrs, I went with the additional temporal dim since it's cleaner, but I'm open to suggestions.

File: torchgeo/datasets/levircd.py (hunk 4 above)
Comment (adamjstewart) [reply torchgeo/datasets/levircd.py]: Looks like it's a known bug in torchvision: https://github.com/pytorch/vision/issues/4108

I think for now the best strategy is to ignore the issue and hope that torchvision cuts a new release soon. Users who need to use LEVIR-CD+ or VHR-10 can simply download and extract the file themselves until then.

**Comment by adamjstewart** [comment]
I guess I have no preference. Let's see what @calebrob6 says.

**Review by adamjstewart — DISMISSED** [review]
Will wait a bit for @calebrob6 to review but otherwise LGTM!

**Comment by calebrob6** [comment]
This looks great! I have a slight preference for `(2, c, h, w)` because it makes what is happening (that there exists 2 images) a little more explicit.

**Comment by isaaccorley** [comment]
Just changed output image to be (t, c, h, w) and made a test for it.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**